Add new stages and images to the Dockerfile

knelli2 opened the pull request:
## Proposed changes

We are getting to the point where we want users (non-developers) to be able to look at SpECTRE and run a few things. Our current Docker image on DockerHub doesn't make this very easy for people. To alleviate this, I have edited the Dockerfile so we have different images for different uses.

The new images will be different tags of a new DockerHub repo `sxscollaboration/spectre`. The three images are:

1. `sxscollaboration/spectre:dev`:  The bare minimum needed to compile SpECTRE. Good for development purposes.
2. `sxscollaboration/spectre:ci`:  Add extra libraries on top of base which are needed to do testing on GitHub CI
3. `sxscollaboration/spectre:demo`: Adds Paraview and has pre-built executables `ExportTimeDependentCoordinates3D` and `EvolveScalarAdvectionKuzmin2D` so users don't have to worry about compiling SpECTRE in order to run a few test problems.

Currently, only the `dev` and `ci` images are on DockerHub. A future PR will edit the `PostRelease` workflow to push the updated pre-built executables to the `demo` image. Also, the `demo` image will also have the `latest` tag which is Docker's default tag for repos. So after the future PR, the default behavior for

```
docker pull sxscollaboration/spectre
```

will be to pull the `demo` image so people can have a clone of the SpECTRE repo and the pre-built executables to play around with.

Thank you to @nilsdeppe for building and pushing the `dev` and `ci` images so this PR can actually be tested on CI.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
Once this is merged, rebase on the latest develop branch to use the new container for CI on your PR.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Edit Dockerfile with new stages

The three stages (tags) are:

1. dev:  The bare minimum needed to compile SpECTRE. Good for
         development purposes
2. ci:   Add extra libraries on top of base which are needed to do
         testing on GitHub CI
3. demo: Adds Paraview and has pre-built executables
         ExportTimeDependentCoordinates3D and
         EvolveScalarAdvectionKuzmin2D so users don't have to worry
         about compiling SpECTRE in order to run a few test problems

knelli2 committed:
Update CI to use new sxscollaboration/spectre:ci image

knelli2 committed:
Remove container support for include-what-you-use

knelli2 committed:
Remove clang-8 and clang-9 from container

nilsdeppe commented on containers/Dockerfile.buildenv:
```diff
@@ -287,3 +280,103 @@ RUN wget https://downloadmirror.intel.com/684899/sde-external-9.0.0-2021-11-07-l
+    && add-apt-repository 'deb http://apt.llvm.org/focal/ llvm-toolchain-focal-13 main' \
+    && apt-get update -y && apt-get install -y clang-13
+
+# `futures` is needed for `yapf` parallelization on Python 2.
```
I guess this isn't true/needed since we don't have yapf for python2 anymore?

knelli2 replied:
Deleted `futures`, rebuilt the container, and then built unit tests and python bindings with python2 and everything worked fine so I think I can get rid of it.

nilsdeppe reviewed (commented):
LGTM, I've pushed the latest container. Just one comment to test out to simplify maintenance if possible. Squash right away if it works, and if not don't worry about it :)

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -2,18 +2,44 @@
+#
+#    To build `demo`, you must be in $SPECTRE_ROOT and there cannot be a
+#    directory named `build` in $SPECTRE_ROOT because the image will create
+#    this directory.
```
Add `(in the container)`

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -2,18 +2,44 @@
+#
+#    and then to push the `demo` image to DockerHub:
+#
+#   docker push sxscollaboration/spectre:dev
```
`:demo`

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -232,20 +209,16 @@ RUN wget http://github.com/xtensor-stack/xsimd/archive/refs/tags/8.1.0.tar.gz \
              -D CMAKE_INSTALL_PREFIX=/usr/local ../ \
     && make install \
-    && cd /work \
+    && cd ../.. \
```
Why this change?

knelli2 replied:
For some reason, it would error if I left it as `cd /work`. Didn't dig into why

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -309,3 +280,103 @@ RUN wget https://downloadmirror.intel.com/684899/sde-external-9.0.0-2021-11-07-l
+
+
+# We don't inherit from ci because that has lots of unnecessary compilers that
+# we don't need and would only increase the size of the image. We inheret from
```
inherit

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -309,3 +280,103 @@ RUN wget https://downloadmirror.intel.com/684899/sde-external-9.0.0-2021-11-07-l
+# We don't inherit from ci because that has lots of unnecessary compilers that
+# we don't need and would only increase the size of the image. We inheret from
+# the remote image rather than the local dev because it is faster on release
+# CI to just pull the remote dev image and buiild demo rather than having to
```
build

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -309,3 +280,103 @@ RUN wget https://downloadmirror.intel.com/684899/sde-external-9.0.0-2021-11-07-l
+# we don't need and would only increase the size of the image. We inheret from
+# the remote image rather than the local dev because it is faster on release
+# CI to just pull the remote dev image and buiild demo rather than having to
+# build all of dev and demo combined.
```
There's a way to fetch the image and then make sure that Docker uses it as cache when building. Maybe add a note here to try this?

knelli2 replied:
Is that not what this does? Pulls the remote image and uses that as cache and builds `demo` on top of it?

If it isn't, could you refer me to what you had in mind?

nilsvu replied:
You would normally base `demo FROM dev`, not from the remote image that must be pushed before. My understanding is you based on the remote image so the post-release workflow doesn't have to rebuild `dev`. And my comment is: there's a way to pre-fetch `dev` so Docker doesn't need to rebuild it. My suggestion is to add a comment so it's clear this is something we could try, but didn't bother just yet.

knelli2 replied:
Ah ok I understand. I added in a couple lines to the fixup explaining this.

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -309,3 +280,103 @@ RUN wget https://downloadmirror.intel.com/684899/sde-external-9.0.0-2021-11-07-l
+    -D MEMORY_ALLOCATOR=SYSTEM \
+    .. \
+  && make $PARALLEL_MAKE_ARG ExportTimeDependentCoordinates3D \
+  && make $PARALLEL_MAKE_ARG EvolveScalarAdvectionKuzmin2D
```
Also build all-pybindings I think, add bin/python to PYTHONPATH, and `pip install jupyterlab` for demos/tutorials.

nilsvu reviewed (commented):
Nice work, thanks for doing this 👍

knelli2 commented:
Pushed a fixup addressing requested changes.
- Tested out that I was able to run `jupyter lab` inside the demo image and view it in chrome
- Removed `futures` from python2 as it was only needed for py2 yapf (which I got rid of)
- Added a `.dockerignore` file in SPECTRE_ROOT that ignores all directories named `build*` just like `.gitignore`. This way you don't have to worry about your build directories being copied from the docker context. (Also had to add this to formaline)

Edit: @nilsdeppe you'll probably have to rebuild the images and restart CI after I squash

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -309,3 +280,120 @@ RUN wget https://downloadmirror.intel.com/684899/sde-external-9.0.0-2021-11-07-l
 # Remove the apt-get cache in order to reduce image size
 RUN apt-get -y clean
+
+FROM dev as ci
```
You can also move the Intel SDE installation to the CI container

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -309,3 +280,120 @@ RUN wget https://downloadmirror.intel.com/684899/sde-external-9.0.0-2021-11-07-l
+  && rm paraview.tar.gz \
+  && mv ParaView-* /opt/paraview
+
+ENV PYTHONPATH "/usr/local/lib/python3.10/dist-packages"
```
What is this line for? Did you mean to point to /opt/paraview?

knelli2 replied:
No, I did have `/opt/paraview` before, but it messed up cmake finding python when we built spectre, even when I gave it the correct python executable. I think I can remove this line.

nilsvu replied:
Ok installing ParaView pybindings can be tricky. I can give it a try later.

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -309,3 +280,120 @@ RUN wget https://downloadmirror.intel.com/684899/sde-external-9.0.0-2021-11-07-l
+  && make ${PARALLEL_MAKE_ARG} EvolveScalarAdvectionKuzmin2D \
+  && make ${PARALLEL_MAKE_ARG} all-pybindings
+
+RUN apt-get update -y && pip3 --no-cache-dir install jupyterlab
```
No need to update apt

nilsvu commented on containers/Dockerfile.buildenv:
```diff
@@ -300,6 +271,56 @@ RUN wget http://mirror.ctan.org/systems/texlive/tlnet/install-tl-unx.tar.gz \
+RUN apt-get -y clean
+
+
+# Everyhting we need to run tests on CI on GitHub.
```
Everything

knelli2 replied:
Why am I so bad at spelling 😓

nilsvu replied:
https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker 😉

knelli2 replied:
What makes it even worse is that I actually have that extension enabled and still missed it. 😅  (To be fair, it got lost in all of the "incorrect" spellings of library names and bash commands)

nilsdeppe commented:
I'm happy with this. @nilsvu please let us know when @knelli2 can squash and if we should wait for your explicit re-approval post-squash :)

nilsvu commented:
Oh I already approved with the comment above, so feel free to squash and don't wait for me with merging this 👍
